commit ca6da08b57c84b7f6c05db2a68378facca02194c
Author:     Yan Cheng <58191769+yanchengnv@users.noreply.github.com>
AuthorDate: Thu Dec 7 11:30:27 2023 -0500
Commit:     GitHub <noreply@github.com>
CommitDate: Thu Dec 7 11:30:27 2023 -0500

    Fix meta file processing in storage and improve schedule job retrieval (#2193)
    
    * Fix meta file processing in storage and improve schedule job retrieval
    
    * changed update_unfinished_jobs to use one get_jobs_by_status
---
 nvflare/private/fed/server/job_runner.py | 18 +++++++++---------
 1 file changed, 9 insertions(+), 9 deletions(-)

diff --git a/nvflare/private/fed/server/job_runner.py b/nvflare/private/fed/server/job_runner.py
index 3d123abc..fd025014 100644
--- a/nvflare/private/fed/server/job_runner.py
+++ b/nvflare/private/fed/server/job_runner.py
@@ -381,10 +381,16 @@ class JobRunner(FLComponent):
             thread.start()
 
             while not self.ask_to_stop:
+                time.sleep(1.0)
+
                 if not isinstance(engine.server.server_state, HotState):
-                    time.sleep(1.0)
                     continue
-                approved_jobs = job_manager.get_jobs_by_status(RunStatus.SUBMITTED, fl_ctx)
+
+                if not engine.get_clients():
+                    # no clients registered yet - don't try to schedule!
+                    continue
+
+                approved_jobs = job_manager.get_jobs_to_schedule(fl_ctx)
                 self.log_debug(
                     fl_ctx, f"{fl_ctx.get_identity_name()} Got approved_jobs: {approved_jobs} from the job_manager"
                 )
@@ -464,8 +470,6 @@ class JobRunner(FLComponent):
                                 fl_ctx, f"Failed to run the Job ({ready_job.job_id}): {secure_format_exception(e)}"
                             )
 
-                time.sleep(1.0)
-
             thread.join()
         else:
             self.log_error(fl_ctx, "There's no Job Manager defined. Won't be able to run the jobs.")
@@ -497,11 +501,7 @@ class JobRunner(FLComponent):
     def update_unfinished_jobs(self, fl_ctx: FLContext):
         engine = fl_ctx.get_engine()
         job_manager = engine.get_component(SystemComponents.JOB_MANAGER)
-        all_jobs = []
-        dispatched_jobs = job_manager.get_jobs_by_status(RunStatus.DISPATCHED, fl_ctx)
-        all_jobs.extend(dispatched_jobs)
-        running_jobs = job_manager.get_jobs_by_status(RunStatus.RUNNING, fl_ctx)
-        all_jobs.extend(running_jobs)
+        all_jobs = job_manager.get_jobs_by_status([RunStatus.RUNNING, RunStatus.DISPATCHED], fl_ctx)
 
         for job in all_jobs:
             try:
